GH-8691: Support age FTP, SMB over time filters

AdamaSorho opened the pull request:
fixes #8691 

Add `LastModifiedFTPFileListFilter` and `LastModifiedSmbFileListFilter` respectively for FTP and SMB to enhance available files filtering based on a last-modified strategy.

AdamaSorho committed:
GH-8691: Support age FTP, SMB over time filters

AdamaSorho committed:
GH-8691: Support age FTP, SMB over time filters

AdamaSorho commented:
I ran the tests before starting to work on the issue, but I am getting an error for `sftp`. So, I did not touch this module. Let me know if I must configure something to pass this test. I will add age support filtering to `sftp` module once I pass the tests. 

See screenshot below:
<img width="1263" alt="SFTP test error" src="https://github.com/spring-projects/spring-integration/assets/37632838/cad6a938-6f25-48b7-a328-dbfc6960f8c9">

artembilan commented:
Thank you for reporting that @AdamaSorho !

Please, consider to add `testRuntimeOnly 'net.i2p.crypto:eddsa:0.3.0'` into `spring-integration-sftp` module. See `build.gradle`.
As part of your PR is OK. Turned out you are not only one who faced this problem for non-default `known_hosts` file.

Also: please, stop posting screenshots for stacktraces: it would much easier for me to copy/paste info for search 😄

AdamaSorho commented:
Understood, sorry about that.

The tests passed now for `sftp` module. Could you tell me why we need `net.i2p.crypto:eddsa:0.3.0`? I don't know understand 🙈

Thank you

AdamaSorho committed:
GH-8691: Support age for sftp module

AdamaSorho committed:
GH-8691: Support age for sftp module

artembilan commented:
We need that library to be able to parse that Ed25519 encrypted key in your known_hosts file: https://github.com/str4d/ed25519-java

AdamaSorho commented:
Ok I see. Thank you

artembilan commented on spring-integration-smb/src/main/java/org/springframework/integration/smb/filters/LastModifiedSmbFileListFilter.java:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	private boolean fileIsAged(SmbFile file, long now) {
+		return file.getLastModified() / ONE_SECOND + this.age <= now;
```
This SMB API does not provide an `Instant` representation, so, we have to stick with millis.
However those blank lines around members an removal of `volatile` are still valid concerns.

artembilan commented on spring-integration-smb/src/test/java/org/springframework/integration/smb/filters/LastModifiedSmbFileListFilterTests.java:
```diff
@@ -0,0 +1,76 @@
+ */
+@SpringJUnitConfig
+@DirtiesContext
+public class LastModifiedSmbFileListFilterTests extends SmbTestSupport {
```
And this one is more drastic than others: it starts a Docker container for Samba server.
Therefore I'd prefer to have all these new tests against mocks and fully in-memory.

AdamaSorho commented on spring-integration-ftp/src/main/java/org/springframework/integration/ftp/filters/LastModifiedFTPFileListFilter.java:
```diff
@@ -0,0 +1,139 @@
+ *
+ * @since 6.2
+ */
+public class LastModifiedFTPFileListFilter implements DiscardAwareFileListFilter<FTPFile> {
```
Only this should be renamed or the other classes too?

artembilan replied:
Right. Others, too. Just follow same pattern for consistency.

AdamaSorho committed:
GH-8691: update tests and class names

AdamaSorho committed:
GH-8691: update tests and class names

artembilan commented on spring-integration-ftp/src/main/java/org/springframework/integration/ftp/filters/FtpLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,147 @@
+	 * @param age the age in seconds.
+	 * @param unit the timeUnit.
+	 */
+	public void setAge(long age, TimeUnit unit) {
```
I don't think we need a `TimeUnit`-based API since we have a `Duration`-one. It really covers any possible time units.

artembilan commented on spring-integration-ftp/src/test/java/org/springframework/integration/ftp/filters/FtpLastModifiedFileListFilterTests.java:
```diff
@@ -0,0 +1,56 @@
+	@Test
+	public void testAge() {
+		FtpLastModifiedFileListFilter filter = new FtpLastModifiedFileListFilter();
+		filter.setAge(60, TimeUnit.SECONDS);
```
Why do we need to set this value if it is exactly what we have by default?

artembilan commented on spring-integration-ftp/src/main/java/org/springframework/integration/ftp/filters/FtpLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2015-2023 the original author or authors.
```
This class was really created in the `2023`. Therefore copyright is just for the current year.

artembilan commented on spring-integration-sftp/src/main/java/org/springframework/integration/sftp/filters/SftpLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2013-2023 the original author or authors.
```
DITTO

artembilan commented on spring-integration-sftp/src/main/java/org/springframework/integration/sftp/filters/SftpLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,143 @@
+	 * @param age the age in seconds.
+	 * @param unit the timeUnit.
+	 */
+	public void setAge(long age, TimeUnit unit) {
```
DITTO

artembilan reviewed (changes requested):
This is very close to the merge.

Now, let's consider to mentioned these new filters in the respective  FTP, SFTP and SMB chapters of the docs.
The change could make it into the `ftp/inbound.adoc`, `sftp/inbound.adoc`, and `smb.adoc`, respectively.

The `whats-new.adoc` also must have a new simple section like `=== Remote Files Support Changes` and mention all new filter there.

Thank you!

AdamaSorho committed:
GH-8691: Remove setAge with TimeUnit

Turned out we don't need anymore since we're using Duration for age in FtpLastModifiedFileListFilter and SftpLastModifiedFileListFilter.

AdamaSorho committed:
GH-8691: Add changes to the docs

artembilan reviewed (changes requested):
Looks like something is wrong with your Git workflow.

1. Switch to `main`
2. Perform `git pull upstream main`
3. Switch back to your branch
4. Perform `git rebase main`
5. `git push origin -f`

Otherwise your PR contains changes you didn't do. Plus it suggests them as new changes somehow...

artembilan commented on spring-integration-smb/src/main/java/org/springframework/integration/smb/filters/SmbLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,142 @@
+	 * @param age the age in seconds.
+	 * @param unit the timeUnit.
+	 */
+	public void setAge(long age, TimeUnit unit) {
```
I believe this one is still has to be remove: the `Duration` does the trick for us anyway.

artembilan reviewed (approved):
OK. Just one nit-pick.
Will take care about that on merge.

Forgot to mention: when we `force push` to the PR, GitHub doesn't notify us about changes.
So, consider in the future to leave some comment that it ready for review.

Thanks

artembilan commented on src/reference/antora/modules/ROOT/pages/ftp/inbound.adoc:
```diff
@@ -98,6 +98,23 @@ You should also understand that the FTP inbound channel adapter is a polling con
+public FtpLastModifiedFileListFilter ftpLastModifiedFileListFilter() {
+	FtpLastModifiedFileListFilter filter = new FtpLastModifiedFileListFilter();
+
+	filter.setAge(Duration.ofSeconds(180));
```
We have ctor for this. Therefore this sample would end up with a single line.
More over I think we don't this examples at all: the API of the class is pretty straightforward and there is a good Javadoc.
We may say though, please, look into its Javadoc for more info.

artembilan reviewed (changes requested):
OK. After looking into this closely, I suggest this evolution.

How about extracting an `AstractLastModifiedFileListFilter<F>` with all the common logic and only leave something like `abstract Instant getLastModified(F remoteFile);` for the implementors?
I believe this way even that existing `LastModifiedFileListFilter` could benefit from new super class.

Yes, some of the implementors would need to use `Instant.ofEpochSecond()` to convert their `long` to expected format by super class.

AdamaSorho commented:
Yeah, that will be perfect!

AdamaSorho committed:
GH-8691: Introduce AbstractLastModifiedFileListFilter

AdamaSorho commented on spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractLastModifiedFileListFilter.java:
```diff
@@ -0,0 +1,124 @@
+
+	protected abstract Instant getLastModified(F remoteFile);
+
+	protected Duration getCurrentAge() {
```
I added this method because we have a `long getAge()` in `LastModifiedFileListFilter` for backward compatibility.

I think we should deprecate `long getAge()`.

AdamaSorho commented on spring-integration-file/src/main/java/org/springframework/integration/file/filters/LastModifiedFileListFilter.java:
```diff
@@ -74,74 +66,16 @@ public LastModifiedFileListFilter(long age) {
+		setAge(unit.toSeconds(age));
 	}
 
 	public long getAge() {
```
We should deprecate this.

artembilan reviewed (changes requested):
This looks very cool!
Just a couple minor comments.

You may not be agreed with me on some of them, so just comment with your arguments.
Otherwise I believe it can be merged after those fixes.

AdamaSorho commented on spring-integration-file/src/main/java/org/springframework/integration/file/filters/LastModifiedFileListFilter.java:
```diff
@@ -74,74 +66,16 @@ public LastModifiedFileListFilter(long age) {
 	 * @param unit the timeUnit.
 	 */
 	public void setAge(long age, TimeUnit unit) {
```
Okay! I think we can remove `..(long age)` constructors and deprecate `LastModifiedFileListFilter(long age)` in favor of `Duration` constructors as well.

artembilan replied:
No, `long` ctor, as well as setter, are good . They are pretty simple convenient. I just don’t see a reason in a `TimeUnit`-based.

AdamaSorho replied:
Okay

AdamaSorho committed:
GH-8691: Deprecate unnecessary methods

artembilan reviewed (approved):
Pulling locally for the final review, clean up and merge...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/73ed3eeebd5b919df852bc484af9a40dc0281307 after some clean up.

Thank you again for contributions!

AdamaSorho commented:
Thank you!

Don't hesitate to ping me on any issue at anytime.
